### DISCUSSION TITLE
Appium v2.0.0-beta.20+

### DISCUSSION THREAD

**Discussion by boneskull** [appium/appium#16148] [Release]
Apologies this isn't especially timely, but just wanted to give a heads-up anyhow.

Besides the usual round of bugfixes, [Appium `v2.0.0-beta.20`](https://github.com/appium/appium/blob/fa01047fba44e7f57ae359ed02d9d5f25a8539bf/packages/appium/CHANGELOG.md#200-beta20-2021-11-15) introduces a major new feature (and associated breaking changes).

## Configuration Files

Appium now supports _configuration files_ ([technical overview](https://github.com/appium/appium/blob/2.0/docs/en/advanced-concepts/configuration-system.md)).  Config files can be used in place of--or in addition to--command-line arguments.

Configuration can be provided in JSON, YAML, or JS (not ESM) format; it can also be embedded in a `package.json`.  For automatic detection, a configuration file should live in the root of your project (e.g., alongside `package.json`) and be named `.appiumrc.<ext>` where `<ext>` is `json`, `yaml`, `yml`, or `js` (see the [overview](https://github.com/appium/appium/blob/2.0/docs/en/advanced-concepts/configuration-system.md) for more details).  Additionally, if `appium` is installed globally, the CLI argument `--config=</path/to/appiumrc>` can be provided to `appium` or `appium server` to specify its location.

The "shape" of the configuration file roughly aligns with the CLI arguments.  We've provided a [JSON schema](https://github.com/appium/appium/blob/2.0/packages/appium/lib/appium-config.schema.json) and some [samples](https://github.com/appium/appium/tree/2.0/sample-code) (see `appium.config.sample.*`) in lieu of proper documentation for now.

Any CLI arguments provided will take precedence over a configuration file, so it's especially good for defining "defaults."

## Breaking Changes: Arguments for Extension Authors

To leverage the configuration system, an extension must provide a `schema` property in its `appium` property of its `package.json`.  The value can be a path to a JSON Schema (either `.js` or `.json`) _or_ a JSON schema itself.  For now, please read the [technical overview](https://github.com/appium/appium/blob/2.0/docs/en/advanced-concepts/configuration-system.md) for more details, or see [appium-xcuitest-driver](https://github.com/appium/appium-xcuitest-driver/blob/master/package.json#L34) for an example.

The `argsConstraints` property of your main class can be removed. The main class will receive any user-provided values from the CLI or config file the same way it did before (instance property `cliArgs`).  `argsConstraints` is no longer supported.

### Renamed Arguments

`--drivers` becomes `--use-drivers` and `--plugins` becomes `--use-plugins`.

### Removed Arguments

`--appium-home` has been removed (use the `APPIUM_HOME` env var instead), as has `--log-filters`.  The latter may be re-added if needed.

### CLI Arguments for Drivers

In Appium v2.0 previous to `beta.20`, some CLI arguments moved into an arg called `--driver-args`.  **`--driver-args` no longer exists.** Instead, installed drivers "contribute" CLI arguments to the `appium` executable.  When present, these will be named `--driver-<driver-name>-<argument-name>` and will be shown when running `appium server --help`.

While still rather verbose, this provides a better experience than providing a JSON string on the CLI!

These arguments are _also_ supported in config files, e.g.,:

```yaml
server:
  driver:
    some-driver:
      some-arg: 'some-value'
```

The above corresponds to `--driver-some-driver-some-arg='some-value'` on the CLI.

Note that there may be some drivers supported in Appium v2.0 which need updating to leverage this new system (though I don't know of any).

Everything described here _also_ goes similarly for plugins, though (as of now) they must be enabled via `--use-plugins` to be shown in `--help`.

**Comment by RSM-SBA** [comment]
I tried to use a simple configuration like in the document (adding server port) and it seems to work when starting the appium server manually. But when I use the AppiumServerBuilder to start the server programmatically, these configurations are not used.
Is that intended or at least known ?

**Reply by KazuCocoa** [reply]
Afaik, client side's launching appium server command does not support Appium 2.0 yet.

**Reply by RSM-SBA** [reply]
Hmm ok, in general it is luckily still working with Appium 2.0 
Would be great if this will not end up in a breaking change because thats a very useful feature to launch the server from code.

**Reply by KazuCocoa** [reply]
It depends on each client development, but probably they will update their code around appium 2.0 rc or after the initial stable release.

**Reply by boneskull** [reply]
Yes, sorry, I'm unfamiliar with `AppiumServerBuilder`.  Which client is this?

**Reply by KazuCocoa** [reply]
I assume java or dotnet client. (I don't know their upcoming version (selenium v4 base) also still have this builder class)

**Reply by RSM-SBA** [reply]
yes its the java client that i use

**Reply by jlipps** [reply]
yes, as @KazuCocoa mentioned, the short answer is that the clients will need to go through and make sure all their extraneous features including subprocess execution are made compatible with Appium 2.0 before its launch. meanwhile, thanks for reporting this, and it would be good to create an issue on the java client repo so that the people who work on that repo make sure to address this particular change.

**Reply by boneskull** [reply]
@jlipps I'd create an issue in there, but I'm not sure what that issue should say.
